specs/math.txt

```diff
@@ -71,57 +82,97 @@ $\$$
+
+```````````````````````````````` example
+$x $ x$
+
+$$ $ $$
+
+$$ $$ $$
+.
+<p>$x $ x$</p>
+<p><span class="math display"> $ </span></p>
+<p><span class="math display"> </span> $$</p>
 ````````````````````````````````
 
-Inline math text with backticks ``$`...`$``:
+Inline math expressions cannot start or end with whitespace, including newlines:
 
 ```````````````````````````````` example
-this is math inline $`y=x`$ test.
-dollar can be contained: $`foo$bar`$
+these are not math texts: $ y=x$, $y=x $, $
+y=x$ and $y=x
+$
 .
-<p>this is math inline <span class="math inline">y=x</span> test.
-dollar can be contained: <span class="math inline">foo$bar</span></p>
+<p>these are not math texts: $ y=x$, $y=x $, $
+y=x$ and $y=x
+$</p>
 ````````````````````````````````
 
-Math expression without backtick cannot start/end with spaces but Math expression with backtick can:
+Inline math expressions do not need to be surrounded with whitespace:
 
 ```````````````````````````````` example
-these are not math texts: $ y=x$ and $y=x $ and $ y=x $
-but these are math texts: $` y=x`$ and $`y=x `$ and $` y=x `$
+these are math texts: foo$y=x$bar and $y=x$bar and foo$y=x$ bar
 .
-<p>these are not math texts: $ y=x$ and $y=x $ and $ y=x $
-but these are math texts: <span class="math inline"> y=x</span> and <span class="math inline">y=x </span> and <span class="math inline"> y=x </span></p>
+<p>these are math texts: foo<span class="math inline">y=x</span>bar and <span class="math inline">y=x</span>bar and foo<span class="math inline">y=x</span> bar</p>
 ````````````````````````````````
 
-Math expression without backtick surrounded with whitespaces:
+Math expressions can be immediately followed by other math expressions:
 
 ```````````````````````````````` example
-these are not math texts: foo$y=x$bar and $y=x$bar and foo$y=x$ bar
-but these are math texts: foo$`y=x`$bar and $`y=x`$bar and foo$`y=x`$ bar
+$a$$b$
+
+$a$$$b$$
+
+$$a$$$b$
+
+$$a$$$$b$$
 .
-<p>these are not math texts: foo$y=x$bar and $y=x$bar and foo$y=x$ bar
-but these are math texts: foo<span class="math inline">y=x</span>bar and <span class="math inline">y=x</span>bar and foo<span class="math inline">y=x</span> bar</p>
+<p><span class="math inline">a</span><span class="math inline">b</span></p>
+<p><span class="math inline">a</span><span class="math display">b</span></p>
+<p><span class="math display">a</span><span class="math inline">b</span></p>
+<p><span class="math display">a</span><span class="math display">b</span></p>
 ````````````````````````````````
 
-Math expression without backtick can be end with punctuation but cannot start with:
+Both inline and display mode math expressions are inline elements with the same
+precedence as code spans. The leftmost valid element takes priority:
 
 ```````````````````````````````` example
-math texts: $x=y$! and $x=y$? and $x=y$: and $x=y$.
-not math texts: !$x=y$! and ?x=y$? and :$x=y$: and .$x=y$.
```

Comment:
This test case should not be removed to follow GitHub's renderer behavior. This PR's implementation parses `$foo$` in `.$foo$` as math expression but Gist's renderer doesn't. It parses entire `.$foo$` as normal text.

Reply:
I agree that the behavior should be tested, but the change to simpler surrounding rules based on the ones from `commonmark-hs` was intentional.

As mentioned, I feel like GitHub's current rules are not so consistent and useful that we really want to replicate them exactly, and I'm afraid they might change in the future. I'm inclined to support things like preceding quotes `"$x$ is even"`, external commas `$a$,$b$ and $c$` and external (en) dashes `the range $a$–$b$`. These currently render as

"$x$ is even"
$a$,$b$ and $c$
the range $a$–$b$